### ISSUE TITLE
Unresolved attribute `test` on type `def a(func: (...) -> Unknown) -> Unknown`

### ISSUE THREAD

**Issue by benglewis** [astral-sh/ty#1256]
### Summary

You should be able to extend callable instances (e.g. functions) with attributes. Currently instead I get the above error for the following code:

```python
def a():
    a.test = 1
```

### Version

ty 0.0.1-alpha.21 (ef52a1940 2025-09-19)

**Comment by carljm** [comment]
Thanks for the report!

I don't think it would be consistent for this to be allowed. In general, statically typed Python assumes that classes have a certain known set of attributes (with known types), and only those attributes can be set (to the correct types) on instances of those classes. Functions do not generally have a `test` attribute.

All major type checkers (mypy, pyright, pyrefly) agree with us that this code is an error. (In mypy you have to add `-> None` to the function definition, otherwise the function body isn't checked at all.)

It's true that setting arbitrary attributes on functions works at runtime, but so does this, and no type checker allows this either:

```py
class A(): pass

a = A()
a.test = 1
```

**Comment by benglewis** [comment]
Weird, because in our existing type checker, basedpyright, I am sure that this is allowed. I will check with regular old pyright and come back if it also allows this behaviour

**Comment by AlexWaygood** [comment]
I have wondered in the past whether we should add a `__setattr__` method to `FunctionType` in typeshed... I do agree that it's quite common (and quite idiomatic!) to monkey-patch attributes onto functions, and that the status quo makes typing many decorators quite painful

**Comment by erictraut** [comment]
In pyright, this is configurable. Pyright has a specific diagnostic category for this check: `reportFunctionMemberAccess`. This check is [disabled](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-settings-defaults) in "basic" type checking mode but is enabled in "standard" and "strict".

**Comment by carljm** [comment]
I'll reopen this, it seems like at some point we may want to consider how we can better support this use case.

**Comment by sharkdp** [comment]
We discussed this internally and we currently do *not* plan any special handling for setting attributes on functions (neither in ty, nor in typeshed).

We understand that this comes up when writing decorators. So while ty will force developers to use suppression comments or `@no_type_check` for the *implementation* of the decorator itself, there is a way to type-annotate the external *signature* of the decorator in such a way that the attribute can be accessed on decorated functions. This makes use of [ty's intersection types](https://docs.astral.sh/ty/features/type-system/#intersection-types) (`Callable[…] & HasAttribute`, where `HasAttribute` is a protocol that describes the existence of the attribute):

```py
from typing import Protocol, Callable, no_type_check
from ty_extensions import Intersection

class HasAttribute(Protocol):
    some_attribute: bytes

@no_type_check
def my_decorator[**P, R](f: Callable[P, R]) -> Intersection[Callable[P, R], HasAttribute]:
    f.some_attribute = 1
    return f

@my_decorator
def g(x: int) -> str:
    raise NotImplementedError

# Calling `g` works fine:
reveal_type(g(1))  # str

# Accessing `some_attribute` works fine:
reveal_type(g.some_attribute)  # bytes

# Accessing an unknown attribute leads to an error:
g.unknown_attribute  # error: unresolved-attribute
```
https://play.ty.dev/bbd2b757-e8f5-472a-91f7-ebfb75010f1a


Please feel free to comment here on this thread if you disagree with our decision, or if you see other use cases that can not be handled this way. We *might* consider using a different rule code for attribute accesses on `FunctionType`, similar to what pyright does.
